### ISSUE TITLE
Disk files cleanup on Teamcity causes problems with file system watching on Linux

### ISSUE THREAD

**Issue by wolfs** [gradle/gradle#19250]
After re-enabling Swabra on CI, there were a lot of file not found exceptions when running ktlint on the build logic sources:
```
A failure occurred while executing org.jlleitschuh.gradle.ktlint.worker.KtLintWorkAction
> /mnt/tcagent1/work/checkout-dir/buildSrc/src/main/kotlin/com/gradle/some/name/MyClass.kt (No such file or directory)
```

It seems like there are also problems on Linux with the way disk files cleanup (Swabra) is handled on Teamcity. See #17865 for a problem we already fixed on Windows.

---
cc: @gradle/execution

**Comment by wolfs** [comment]
I think I found the root cause.

So we already knew that the disk file cleaner on Teamcity (aka Swabra) moves away the directory it wants to clean up before cleaning it up. Though it also seems like the cleanup doesn't happen instantly after moving the directory. Instead, Swabra schedules a job which removes those directories in the background.

So I think this is what happened:
- A build did run which started watching buildSrc/.
- The next build decided that it needs a clean checkout (different branch). So it moves the old checkout directory to `/home/tcagent1/agent/work/.old/<checkout-dir-name>_<counter>`, *without deleting it*
- Since the deletion of the checkout dir did not happen, Gradle does not receive any file system events from it.
- Teamcity checks out the code in the checkout dir location and runs a build. Gradle assumes it is watching the sources, although it doesn't. So it uses the snapshot from memory which has the wrong files in it.
- When Swabra finally deletes the moved away checkout directory, Gradle receives the deletion events and stops watching the location.

The file system watching probe didn't detect the problem, since it was fired in the location before the move happened, and will arrive in the new location with inotify still receiving the event. For the next build the event won't come back, since the event will be fired in the new location and the new location is not watched.

**Comment by wolfs** [comment]
I can think of some solutions:
1. We also check on Linux whether the directories we watch still are at the same path. I didn't find an API for doing this, though.
2. We could watch all parent directories of watched directories. So we would be informed if one of the directories is moved.
3. We could watch the watchable hierarchy roots as well, so we'd be notified if the whole hierarchy is moved.

It seems like 3. would be the cheapest option which would fix the problem.

**Comment by wolfs** [comment]
We decided to go with 1 and 3:
- When registering a watch point on Linux, we'd record the inode. Then we'd check at the beginning of the build if the inode of path we watch is still the same, and if it isn't we stop watching the location. This is similar to what we do on Windows now.
- We still are going to watch the roots of the watchable hierarchies, even for non-hierarchical watchers.

**Comment by wolfs** [comment]
While #19255 fixes the issue, the full fix is #19278.

**Comment by wolfs** [comment]
Fixed via https://github.com/gradle/gradle/pull/19278.
